core/src/server/subscription.rs

```diff
@@ -0,0 +1,475 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Subscription related types and traits for server implementations.
+
+use super::helpers::{MethodResponse, MethodSink};
+use crate::server::error::{DisconnectError, SendTimeoutError, SubscriptionAcceptRejectError, TrySendError};
+use crate::server::rpc_module::ConnectionId;
+use crate::{traits::IdProvider, Error};
+use jsonrpsee_types::{response::SubscriptionError, ErrorObjectOwned, Id, SubscriptionId, SubscriptionResponse};
+use parking_lot::Mutex;
+use rustc_hash::FxHashMap;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{sync::Arc, time::Duration};
+use tokio::sync::{mpsc, Notify, OwnedSemaphorePermit, Semaphore};
+
+/// Type-alias for subscribers.
+pub type Subscribers = Arc<Mutex<FxHashMap<SubscriptionKey, (MethodSink, mpsc::Receiver<()>)>>>;
+
+/// Convert something into a subscription close notification
+/// before a subscription is terminated.
+pub trait IntoSubscriptionResponse {
+	/// Convert something into a subscription response
+	fn into_response(self) -> SubscriptionCloseResponse;
+}
+
+/// Represents what action that will sent when a subscription callback returns.
+#[derive(Debug)]
+pub enum SubscriptionCloseResponse {
+	/// No further message will be sent.
+	None,
+	/// Send a ordinary subscription response.
+	Some(SubscriptionMessage),
+	/// Send a subscription error response.
+	Err(SubscriptionMessage),
+}
+
+impl<T> IntoSubscriptionResponse for Option<T>
```

Comment:
Nit: I feel like this name, `IntoSubscriptionResponse`, should be consistent with the name of the thing that's it converts stuff into (ie maybe `SubscriptionResponse` rather than `SubscriptionCloseResponse`)